### PR TITLE
[93] create unified tiering policy engine (master)

### PR DESCRIPTION
**Pull request by jasoncoposky**

### CODE REVIEW

File: example_unified_tiering_invocation.r [hunk 1]
```diff
@@ -0,0 +1,11 @@
+{
+   "rule-engine-instance-name": "irods_rule_engine_plugin-unified_storage_tiering-instance",
+   "rule-engine-operation": "irods_policy_schedule_storage_tiering",
+   "delay-parameters": "<INST_NAME>irods_rule_engine_plugin-unified_storage_tiering-instance</INST_NAME><PLUSET>1s</PLUSET><EF>1h DOUBLE UNTIL SUCCESS OR 6 TIMES</EF>",
+   "storage-tier-groups": [
+       "example_group_g2",
```
Comment (trel) [inline example_unified_tiering_invocation.r]: you want both groups here by default?

File: libirods_rule_engine_plugin-unified_storage_tiering.cpp [hunk 2]
```diff
@@ -0,0 +1,836 @@
+// nclude "irods_re_ruleexistshelper.hpp"
```
Comment (trel) [inline libirods_rule_engine_plugin-unified_storage_tiering.cpp]: something seems off, here.

File: unified_storage_tiering.cmake [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+set(CPACK_DEBIAN_${IRODS_PACKAGE_COMPONENT_POLICY_NAME_UPPERCASE}_PACKAGE_DEPENDS "${IRODS_PACKAGE_DEPENDENCIES_STRING}, irods-server (= ${IRODS_VERSION}), irods-runtime (= ${IRODS_VERSION}), libc6, irods-rule-engine-plugin-apply-access-time-2.5.0, irods-rule-engine-plugin-data-movement-2.5.0, irods-rule-engine-plugin-data-replication-2.5.0, irods-rule-engine-plugin-data-verification-2.5.0")
+
+set(CPACK_RPM_${IRODS_PACKAGE_COMPONENT_POLICY_NAME}_PACKAGE_NAME ${TARGET_NAME})
+if (IRODS_LINUX_DISTRIBUTION_NAME STREQUAL "centos" OR IRODS_LINUX_DISTRIBUTION_NAME STREQUAL "centos linux")
+    set(CPACK_RPM_${IRODS_PACKAGE_COMPONENT_POLICY_NAME}_PACKAGE_REQUIRES "${IRODS_PACKAGE_DEPENDENCIES_STRING}, irods-server = ${IRODS_VERSION}, irods-runtime = ${IRODS_VERSION}, openssl")
+elseif (IRODS_LINUX_DISTRIBUTION_NAME STREQUAL "opensuse")
+    set(CPACK_RPM_${IRODS_PACKAGE_COMPONENT_POLICY_NAME}_PACKAGE_REQUIRES "${IRODS_PACKAGE_DEPENDENCIES_STRING}, irods-server = ${IRODS_VERSION}, irods-runtime = ${IRODS_VERSION}, libopenssl1_0_0")
+endif()
+
+set(CPACK_DEBIAN_PACKAGE_BREAKS "irods-rule-engine-plugin-tiered-storage")
```
Comment (trel) [inline unified_storage_tiering.cmake]: these three lines aren't correct - this unified one, i don't think, will be replacing the misnamed first release of this plugin...   discush...

File: example_unified_tiering_invocation.r (hunk 1 above)
Comment (jasoncoposky) [reply example_unified_tiering_invocation.r]: yes, this is for testing.

File: libirods_rule_engine_plugin-unified_storage_tiering.cpp (hunk 2 above)
Comment (jasoncoposky) [reply libirods_rule_engine_plugin-unified_storage_tiering.cpp]: just cruft

File: unified_storage_tiering.cmake (hunk 3 above)
Comment (jasoncoposky) [reply unified_storage_tiering.cmake]: true, left from copy-paste.

File: example_unified_tiering_invocation.r (hunk 1 above)
Comment (trel) [reply example_unified_tiering_invocation.r]: gotcha

File: unified_storage_tiering.cmake (hunk 3 above)
Comment (trel) [reply unified_storage_tiering.cmake]: cool - please to torch.

File: example_unified_tiering_invocation.r (hunk 1 above)
Comment (jasoncoposky) [reply example_unified_tiering_invocation.r]: wrt apply_ vs update_, the collections themselves are not annotated with access time just the objects.  we are not updating an access time for a collection.

**Comment by trel** [comment]
looks good.... tests pass?   will wait for kory vote...

**Comment by korydraughn** [comment]
Everything looks fine as far as continuation is concerned. I'd recommend waiting until the REPF changes are in so this can be tested against it. The REPF changes are being tested now.

File: libirods_rule_engine_plugin-unified_storage_tiering.cpp [hunk 4]
```diff
@@ -0,0 +1,835 @@
...
+    irods::callback        _eff_hdlr) {
+    using json = nlohmann::json;
+
+    try {
+        // skip the first line: @external
+        std::string rule_text{_rule_text};
+        if(_rule_text.find("@external") != std::string::npos) {
+            rule_text = _rule_text.substr(10);
+        }
+        const auto rule_obj = json::parse(rule_text);
```
Comment (korydraughn) [inline libirods_rule_engine_plugin-unified_storage_tiering.cpp]: This can throw `nlohmann::json::parse_error`. Missing catch.

**Pull request merged**